<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Stijn Hawinkel" />


<title>Vignette of the oosse package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Vignette of the oosse package</h1>
<h4 class="author">Stijn Hawinkel</h4>


<div id="TOC">
<ul>
<li><a href="#introduction" id="toc-introduction"><span class="toc-section-number">1</span> Introduction</a></li>
<li><a href="#installation-instructions" id="toc-installation-instructions"><span class="toc-section-number">2</span> Installation instructions</a></li>
<li><a href="#illustrative-examples" id="toc-illustrative-examples"><span class="toc-section-number">3</span>
Illustrative examples</a>
<ul>
<li><a href="#linear-model" id="toc-linear-model"><span class="toc-section-number">3.1</span> Linear model</a></li>
<li><a href="#regularised-linear-model" id="toc-regularised-linear-model"><span class="toc-section-number">3.2</span> Regularised linear model</a></li>
<li><a href="#random-forest" id="toc-random-forest"><span class="toc-section-number">3.3</span> Random forest</a></li>
</ul></li>
<li><a href="#session-info" id="toc-session-info"><span class="toc-section-number">4</span> Session info</a></li>
</ul>
</div>

<div id="introduction" class="section level1" number="1">
<h1><span class="header-section-number">1</span> Introduction</h1>
<p>This vignette demonstrates the use of the <strong>oosse</strong>
package for estimating for estimating out-of-sample R² and its standard
error through resampling algorithms, described in <a href="https://doi.org/10.48550/arXiv.2302.05131">“Out-of-sample R²:
estimation and inference”</a> by Hawinkel et al. 2023.</p>
</div>
<div id="installation-instructions" class="section level1" number="2">
<h1><span class="header-section-number">2</span> Installation
instructions</h1>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">install.packages</span>(<span class="st">&quot;oosse&quot;</span>)</span></code></pre></div>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a><span class="fu">library</span>(oosse)</span></code></pre></div>
</div>
<div id="illustrative-examples" class="section level1" number="3">
<h1><span class="header-section-number">3</span> Illustrative
examples</h1>
<p>The <em>R2oosse</em> function works with any pair of fitting and
prediction functions. Here we illustrate a number of them, but any
prediction function implemented in R can be used. The built-in dataset
<em>Brassica</em> is used, which contains <em>rlog</em>-transformed gene
expression measurements for the 1,000 most expressed genes in the
<em>Expr</em> slot, as well as 5 outcome phenotypes in the
<em>Pheno</em> slot.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a><span class="fu">data</span>(Brassica)</span></code></pre></div>
<div id="linear-model" class="section level2" number="3.1">
<h2><span class="header-section-number">3.1</span> Linear model</h2>
<p>The fitting model must accept at least an outcome vector <em>y</em>
and a regressor matrix <em>x</em>:</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>fitFunLM <span class="ot">=</span> <span class="cf">function</span>(y, x){<span class="fu">lm.fit</span>(<span class="at">y =</span> y, <span class="at">x =</span> <span class="fu">cbind</span>(<span class="dv">1</span>, x))}</span></code></pre></div>
<p>The predictive model must accept arguments <em>mod</em> (the fitted
model) and <em>x</em>, the regressor matrix for a new set of
observations.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a>predFunLM <span class="ot">=</span> <span class="cf">function</span>(mod, x) {<span class="fu">cbind</span>(<span class="dv">1</span>,x) <span class="sc">%*%</span> mod<span class="sc">$</span>coef}</span></code></pre></div>
<p>Now that these functions have been defined, we apply the prediction
model for leaf_8_width using the first 10 genes. Multithreading is used
automatically using the <em>BiocParallel</em> package. Change the
following setup depending on your system.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>nCores <span class="ot">=</span> <span class="dv">2</span> <span class="co"># For CRAN build max 2</span></span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a><span class="fu">library</span>(BiocParallel)</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a><span class="cf">if</span>(.Platform<span class="sc">$</span>OS.type <span class="sc">==</span> <span class="st">&quot;unix&quot;</span>){</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>    <span class="co">#On unix-based systems, use MulticoreParam</span></span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>    <span class="fu">register</span>(<span class="fu">MulticoreParam</span>(nCores))</span>
<span id="cb6-6"><a href="#cb6-6" tabindex="-1"></a>} <span class="cf">else</span> {</span>
<span id="cb6-7"><a href="#cb6-7" tabindex="-1"></a>    <span class="co">#On windows, use makeCluster</span></span>
<span id="cb6-8"><a href="#cb6-8" tabindex="-1"></a>    <span class="fu">library</span>(doParallel)</span>
<span id="cb6-9"><a href="#cb6-9" tabindex="-1"></a>    Clus <span class="ot">=</span> <span class="fu">makeCluster</span>(nCores)</span>
<span id="cb6-10"><a href="#cb6-10" tabindex="-1"></a>    <span class="fu">registerDoParallel</span>(Clus)</span>
<span id="cb6-11"><a href="#cb6-11" tabindex="-1"></a>    <span class="fu">register</span>(<span class="fu">DoparParam</span>(), <span class="at">default =</span> <span class="cn">TRUE</span>)</span>
<span id="cb6-12"><a href="#cb6-12" tabindex="-1"></a>}</span></code></pre></div>
<p>Now estimate out-of-sample <span class="math inline">\(R^2\)</span>,
also a rough estimate of the computation time is given. Remember to
provide the cluster for multithreading.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>R2lm <span class="ot">=</span> <span class="fu">R2oosse</span>(<span class="at">y =</span> Brassica<span class="sc">$</span>Pheno<span class="sc">$</span>Leaf_8_width, <span class="at">x =</span> Brassica<span class="sc">$</span>Expr[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], </span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>               <span class="at">fitFun =</span> fitFunLM, <span class="at">predFun =</span> predFunLM)</span></code></pre></div>
<pre><code>## Fitting and evaluating the model once took 0 seconds.
## You requested 200 repeats of 10-fold cross-validation with 2 cores, which is expected to last for roughly
## 0 seconds</code></pre>
<p>Estimates and standard error of the different components are now
available.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="co">#R2</span></span>
<span id="cb9-2"><a href="#cb9-2" tabindex="-1"></a>R2lm<span class="sc">$</span>R2</span></code></pre></div>
<pre><code>##        R2      R2SE 
## 0.4832506 0.1151979</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" tabindex="-1"></a><span class="co">#MSE</span></span>
<span id="cb11-2"><a href="#cb11-2" tabindex="-1"></a>R2lm<span class="sc">$</span>MSE</span></code></pre></div>
<pre><code>##       MSE     MSESE 
## 2.9554375 0.4792718</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" tabindex="-1"></a><span class="co">#MST</span></span>
<span id="cb13-2"><a href="#cb13-2" tabindex="-1"></a>R2lm<span class="sc">$</span>MST</span></code></pre></div>
<pre><code>##      MST    MSTSE 
## 5.719286 1.035600</code></pre>
<p>Also confidence intervals can be constructed:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb15-1"><a href="#cb15-1" tabindex="-1"></a><span class="co"># R2</span></span>
<span id="cb15-2"><a href="#cb15-2" tabindex="-1"></a><span class="fu">buildConfInt</span>(R2lm)</span></code></pre></div>
<pre><code>##      2.5%     97.5% 
## 0.2574668 0.7090344</code></pre>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" tabindex="-1"></a><span class="co">#MSE, 90% confidence interval</span></span>
<span id="cb17-2"><a href="#cb17-2" tabindex="-1"></a><span class="fu">buildConfInt</span>(R2lm, <span class="at">what =</span> <span class="st">&quot;MSE&quot;</span>, <span class="at">conf =</span> <span class="fl">0.9</span>)</span></code></pre></div>
<pre><code>##       5%      95% 
## 2.167106 3.743769</code></pre>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb19-1"><a href="#cb19-1" tabindex="-1"></a><span class="co">#MST, based on chi-square distribution</span></span>
<span id="cb19-2"><a href="#cb19-2" tabindex="-1"></a><span class="fu">buildConfInt</span>(R2lm, <span class="at">what =</span> <span class="st">&quot;MST&quot;</span>)</span></code></pre></div>
<pre><code>##     2.5%    97.5% 
## 4.129867 8.446729</code></pre>
<p>By default, cross-validation (CV) is used to estimate the MSE, and
nonparametric bootstrapping is used to estimate the correlation between
MSE and MST estimators. Other choices can be supplied though, e.g. for
bootstrap .632 estimation of the MSE and jackknife estimation of the
correlation:</p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb21-1"><a href="#cb21-1" tabindex="-1"></a>R2lm632jn <span class="ot">=</span> <span class="fu">R2oosse</span>(<span class="at">y =</span> Brassica<span class="sc">$</span>Pheno<span class="sc">$</span>Leaf_8_width, <span class="at">x =</span> Brassica<span class="sc">$</span>Expr[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], </span>
<span id="cb21-2"><a href="#cb21-2" tabindex="-1"></a>                    <span class="at">fitFun =</span> fitFunLM, <span class="at">predFun =</span> predFunLM, <span class="at">methodMSE =</span> <span class="st">&quot;bootstrap&quot;</span>,</span>
<span id="cb21-3"><a href="#cb21-3" tabindex="-1"></a>                    <span class="at">methodCor =</span> <span class="st">&quot;jackknife&quot;</span>)</span></code></pre></div>
<pre><code>## Fitting and evaluating the model once took 0 seconds.
## You requested 200 .632 bootstrap instances with 2 cores, which is expected to last for roughly
## 19.2 seconds</code></pre>
<p>Supplying a dataframe with predictor variables is not allowed. The
user is asked to build the design matrix yourself with model.matrix
prior to calling <em>R2oosse</em>:</p>
<div class="sourceCode" id="cb23"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb23-1"><a href="#cb23-1" tabindex="-1"></a><span class="co">#We construct a fake data frame also containing genotypes</span></span>
<span id="cb23-2"><a href="#cb23-2" tabindex="-1"></a>fakeDf <span class="ot">=</span> <span class="fu">data.frame</span>(Brassica<span class="sc">$</span>Expr[, <span class="dv">1</span><span class="sc">:</span><span class="dv">5</span>], <span class="st">&quot;genotype&quot;</span> <span class="ot">=</span> <span class="fu">sample</span>(<span class="fu">c</span>(<span class="st">&quot;Genotype1&quot;</span>, <span class="st">&quot;Genotype2&quot;</span>, <span class="st">&quot;Genotype3&quot;</span>), <span class="at">replace =</span> <span class="cn">TRUE</span>, <span class="fu">nrow</span>(Brassica<span class="sc">$</span>Expr)))</span>
<span id="cb23-3"><a href="#cb23-3" tabindex="-1"></a><span class="co">#Build the design matrix. The model.matrix variables automatically constructs dummy variables</span></span>
<span id="cb23-4"><a href="#cb23-4" tabindex="-1"></a>designMatrix <span class="ot">=</span> <span class="fu">model.matrix</span>(<span class="sc">~</span> .  , <span class="at">data =</span> fakeDf)[, <span class="sc">-</span><span class="dv">1</span>] <span class="co">#Include no intercept as fitting function already does this</span></span>
<span id="cb23-5"><a href="#cb23-5" tabindex="-1"></a><span class="co">#Now run oosse</span></span>
<span id="cb23-6"><a href="#cb23-6" tabindex="-1"></a>R2modMat <span class="ot">=</span> <span class="fu">R2oosse</span>(<span class="at">y =</span> Brassica<span class="sc">$</span>Pheno<span class="sc">$</span>Leaf_8_width, <span class="at">x =</span> designMatrix, </span>
<span id="cb23-7"><a href="#cb23-7" tabindex="-1"></a>                    <span class="at">fitFun =</span> fitFunLM, <span class="at">predFun =</span> predFunLM)</span></code></pre></div>
<pre><code>## Fitting and evaluating the model once took 0 seconds.
## You requested 200 repeats of 10-fold cross-validation with 2 cores, which is expected to last for roughly
## 0 seconds</code></pre>
</div>
<div id="regularised-linear-model" class="section level2" number="3.2">
<h2><span class="header-section-number">3.2</span> Regularised linear
model</h2>
<p>For high-dimensional problems, such as the Brassica dataset, a
regularised linear model is better suited to incorporate information for
all genes. We use the <em>cv.glmnet</em> function from the
<em>glmnet</em> package, which includes internal cross-validation for
tuning the penalty parameter. Following custom function definitions are
needed to fit in with the naming convention of the <em>oosse</em>
package.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb25-1"><a href="#cb25-1" tabindex="-1"></a>fitFunReg <span class="ot">=</span> <span class="cf">function</span>(y, x, ...) {<span class="fu">cv.glmnet</span>(<span class="at">y =</span> y, <span class="at">x =</span> x, ...)}</span>
<span id="cb25-2"><a href="#cb25-2" tabindex="-1"></a>predFunReg <span class="ot">=</span> <span class="cf">function</span>(mod, x, ...){<span class="fu">predict</span>(mod, <span class="at">newx =</span> x, ...)}</span></code></pre></div>
<p>We adapt the parameter settings a bit to reduce computation time of
the vignette, it is recommended to use 10-fold cross-validation, at
least 200 repeats of the cross-validation splits and 50 bootstrap
replicates.</p>
<div class="sourceCode" id="cb26"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb26-1"><a href="#cb26-1" tabindex="-1"></a>nFolds <span class="ot">=</span> <span class="dv">5</span>; cvReps <span class="ot">=</span> <span class="fl">1e2</span>; nBoots <span class="ot">=</span> <span class="fl">4e1</span>;numFeat <span class="ot">=</span> <span class="dv">25</span></span>
<span id="cb26-2"><a href="#cb26-2" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">require</span>(glmnet)){</span>
<span id="cb26-3"><a href="#cb26-3" tabindex="-1"></a>    <span class="cf">if</span>(onWindows <span class="ot">&lt;-</span> (.Platform<span class="sc">$</span>OS.type <span class="sc">==</span> <span class="st">&quot;windows&quot;</span>)){</span>
<span id="cb26-4"><a href="#cb26-4" tabindex="-1"></a>        <span class="fu">clusterEvalQ</span>(Clus, <span class="fu">require</span>(glmnet))</span>
<span id="cb26-5"><a href="#cb26-5" tabindex="-1"></a>    }</span>
<span id="cb26-6"><a href="#cb26-6" tabindex="-1"></a>    R2pen <span class="ot">=</span> <span class="fu">R2oosse</span>(<span class="at">y =</span> Brassica<span class="sc">$</span>Pheno<span class="sc">$</span>Leaf_8_width, <span class="at">x =</span> Brassica<span class="sc">$</span>Expr[, <span class="fu">seq_len</span>(numFeat)], <span class="co">#Subset genes for speed</span></span>
<span id="cb26-7"><a href="#cb26-7" tabindex="-1"></a>                    <span class="at">nFolds =</span> nFolds, <span class="at">cvReps =</span> cvReps, <span class="at">nBootstrapsCor =</span> nBoots,</span>
<span id="cb26-8"><a href="#cb26-8" tabindex="-1"></a>                        <span class="at">fitFun =</span> fitFunReg, <span class="at">predFun =</span> predFunReg, <span class="at">alpha =</span> <span class="dv">1</span>)<span class="co">#Lasso model</span></span>
<span id="cb26-9"><a href="#cb26-9" tabindex="-1"></a>    R2pen<span class="sc">$</span>R2</span>
<span id="cb26-10"><a href="#cb26-10" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## Loading required package: glmnet</code></pre>
<pre><code>## Loading required package: Matrix</code></pre>
<pre><code>## Loaded glmnet 4.1-7</code></pre>
<pre><code>## Fitting and evaluating the model once took 0.06 seconds.
## You requested 100 repeats of 5-fold cross-validation with 2 cores, which is expected to last for roughly
## 1 minutes and 21 seconds</code></pre>
<pre><code>##         R2       R2SE 
## 0.62872596 0.09354877</code></pre>
</div>
<div id="random-forest" class="section level2" number="3.3">
<h2><span class="header-section-number">3.3</span> Random forest</h2>
<p>As a final example we use a random forest as a prediction model. We
use the implementation from the <em>randomForest</em> package.</p>
<div class="sourceCode" id="cb32"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb32-1"><a href="#cb32-1" tabindex="-1"></a><span class="cf">if</span>(<span class="fu">require</span>(randomForest)){</span>
<span id="cb32-2"><a href="#cb32-2" tabindex="-1"></a>   <span class="cf">if</span>(onWindows){</span>
<span id="cb32-3"><a href="#cb32-3" tabindex="-1"></a>        <span class="fu">clusterEvalQ</span>(Clus, <span class="fu">require</span>(randomForest))</span>
<span id="cb32-4"><a href="#cb32-4" tabindex="-1"></a>    }</span>
<span id="cb32-5"><a href="#cb32-5" tabindex="-1"></a>    fitFunrf <span class="ot">=</span> <span class="cf">function</span>(y, x, ...){<span class="fu">randomForest</span>(<span class="at">y =</span> y, x, ...)}</span>
<span id="cb32-6"><a href="#cb32-6" tabindex="-1"></a>    predFunrf <span class="ot">=</span> <span class="cf">function</span>(mod, x, ...){<span class="fu">predict</span>(mod, x, ...)}</span>
<span id="cb32-7"><a href="#cb32-7" tabindex="-1"></a>    R2rf <span class="ot">=</span> <span class="fu">R2oosse</span>(<span class="at">y =</span> Brassica<span class="sc">$</span>Pheno<span class="sc">$</span>Leaf_8_width, <span class="at">x =</span> Brassica<span class="sc">$</span>Expr[, <span class="fu">seq_len</span>(numFeat)],</span>
<span id="cb32-8"><a href="#cb32-8" tabindex="-1"></a>                     <span class="at">nFolds =</span> nFolds, <span class="at">cvReps =</span> cvReps, <span class="at">nBootstrapsCor =</span> nBoots,</span>
<span id="cb32-9"><a href="#cb32-9" tabindex="-1"></a>                        <span class="at">fitFun =</span> fitFunrf, <span class="at">predFun =</span> predFunrf)</span>
<span id="cb32-10"><a href="#cb32-10" tabindex="-1"></a>    R2rf<span class="sc">$</span>R2</span>
<span id="cb32-11"><a href="#cb32-11" tabindex="-1"></a>}</span></code></pre></div>
<pre><code>## Loading required package: randomForest</code></pre>
<pre><code>## randomForest 4.7-1.1</code></pre>
<pre><code>## Type rfNews() to see new features/changes/bug fixes.</code></pre>
<pre><code>## Fitting and evaluating the model once took 0.05 seconds.
## You requested 100 repeats of 5-fold cross-validation with 2 cores, which is expected to last for roughly
## 1 minutes and 8.85 seconds</code></pre>
<pre><code>##        R2      R2SE 
## 0.6981843 0.0816148</code></pre>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb38-1"><a href="#cb38-1" tabindex="-1"></a><span class="cf">if</span>(onWindows){</span>
<span id="cb38-2"><a href="#cb38-2" tabindex="-1"></a>    <span class="fu">stopCluster</span>(Clus)</span>
<span id="cb38-3"><a href="#cb38-3" tabindex="-1"></a>}</span></code></pre></div>
<p>The <span class="math inline">\(R^2\)</span> estimate is comparable
to that of the penalised regression model.</p>
</div>
</div>
<div id="session-info" class="section level1" number="4">
<h1><span class="header-section-number">4</span> Session info</h1>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb39-1"><a href="#cb39-1" tabindex="-1"></a><span class="fu">sessionInfo</span>()</span></code></pre></div>
<pre><code>## R version 4.3.1 (2023-06-16)
## Platform: x86_64-pc-linux-gnu (64-bit)
## Running under: Ubuntu 22.04.3 LTS
## 
## Matrix products: default
## BLAS:   /usr/lib/x86_64-linux-gnu/blas/libblas.so.3.10.0 
## LAPACK: /usr/lib/x86_64-linux-gnu/lapack/liblapack.so.3.10.0
## 
## locale:
##  [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              
##  [3] LC_TIME=de_BE.UTF-8        LC_COLLATE=C              
##  [5] LC_MONETARY=de_BE.UTF-8    LC_MESSAGES=en_US.UTF-8   
##  [7] LC_PAPER=de_BE.UTF-8       LC_NAME=C                 
##  [9] LC_ADDRESS=C               LC_TELEPHONE=C            
## [11] LC_MEASUREMENT=de_BE.UTF-8 LC_IDENTIFICATION=C       
## 
## time zone: Europe/Amsterdam
## tzcode source: system (glibc)
## 
## attached base packages:
## [1] stats     graphics  grDevices utils     datasets  methods   base     
## 
## other attached packages:
## [1] randomForest_4.7-1.1 glmnet_4.1-7         Matrix_1.6-0        
## [4] BiocParallel_1.34.2  oosse_1.0.10        
## 
## loaded via a namespace (and not attached):
##  [1] doParallel_1.0.17 cli_3.6.1         knitr_1.43        rlang_1.1.1      
##  [5] xfun_0.39         jsonlite_1.8.7    htmltools_0.5.5   sass_0.4.7       
##  [9] rmarkdown_2.23    grid_4.3.1        evaluate_0.21     jquerylib_0.1.4  
## [13] fastmap_1.1.1     yaml_2.3.7        foreach_1.5.2     compiler_4.3.1   
## [17] codetools_0.2-19  Rcpp_1.0.11       rstudioapi_0.15.0 lattice_0.21-8   
## [21] digest_0.6.33     R6_2.5.1          splines_4.3.1     shape_1.4.6      
## [25] Rdpack_2.4        parallel_4.3.1    rbibutils_2.2.13  bslib_0.5.0      
## [29] tools_4.3.1       iterators_1.0.14  survival_3.5-5    cachem_1.0.8</code></pre>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
